### ISSUE TITLE
[v2.72.0] Creating Blob fails with 403

### ISSUE THREAD

**Issue by AleksandarDev** [pulumi/pulumi-azure-native#3709]
### What happened?

Getting 403 for Storage Blob create operation with `@pulumi/azure-native 2.72.0` for `pulumi up` (`pulumi preview` works fine).

Error is not present with `@pulumi/azure-native 2.71.0`.

```
azure-native:storage:Blob (zip-rbpub-99bb43827bec0147e5c44a8e1d37fa65):
    error: retrieving blob properties "zip-rbpub-99bb43827bec0147e5c44a8e1d37fa65.zip" (container "zips" / account "safuncs3a899f6a"): HEAD https://safuncs3a899f6a.blob.core.windows.net/zips/zip-rbpub-99bb43827bec0147e5c44a8e1d37fa65.zip
    --------------------------------------------------------------------------------
    RESPONSE 403: 403 This request is not authorized to perform this operation using this permission.
    ERROR CODE: AuthorizationPermissionMismatch
    --------------------------------------------------------------------------------
    Response contained no body
    --------------------------------------------------------------------------------

  pulumi:pulumi:Stack (cloud-primary-next):
    error: update failed
    error: Running program 'D:\source\signalco\infra\apps\cloud-primary\dist\index.js' failed with an unhandled exception:
    <ref *1> Error: invocation of command:local:run returned an error: grpc: the client connection is closing
        at Object.callback (D:\source\signalco\infra\node_modules\.pnpm\@pulumi+pulumi@3.139.0_ts-node@7.0.1_typescript@5.6.3\node_modules\@pulumi\runtime\invoke.ts:275:37)
        at Object.onReceiveStatus (D:\source\signalco\infra\node_modules\.pnpm\@grpc+grpc-js@1.10.1\node_modules\@grpc\grpc-js\src\client.ts:360:26)
        at Object.onReceiveStatus (D:\source\signalco\infra\node_modules\.pnpm\@grpc+grpc-js@1.10.1\node_modules\@grpc\grpc-js\src\client-interceptors.ts:458:34)
        at Object.onReceiveStatus (D:\source\signalco\infra\node_modules\.pnpm\@grpc+grpc-js@1.10.1\node_modules\@grpc\grpc-js\src\client-interceptors.ts:419:48)
        at D:\source\signalco\infra\node_modules\.pnpm\@grpc+grpc-js@1.10.1\node_modules\@grpc\grpc-js\src\resolving-call.ts:132:24
        at processTicksAndRejections (node:internal/process/task_queues:77:11) {
      promise: Promise { <rejected> [Circular *1] }
    }

  azure-native:storage:Blob (zip-cpub-8b4497661bbe3355a5f8722762066a53):
    error: retrieving blob properties "zip-cpub-8b4497661bbe3355a5f8722762066a53.zip" (container "zips" / account "safuncs3a899f6a"): HEAD https://safuncs3a899f6a.blob.core.windows.net/zips/zip-cpub-8b4497661bbe3355a5f8722762066a53.zip
    --------------------------------------------------------------------------------
    RESPONSE 403: 403 This request is not authorized to perform this operation using this permission.
    ERROR CODE: AuthorizationPermissionMismatch
    --------------------------------------------------------------------------------
    Response contained no body
    --------------------------------------------------------------------------------
```

### Example

There is no code changes other than package update.

PR: https://github.com/signalco-io/signalco/pull/6244
Build: https://github.com/signalco-io/signalco/actions/runs/11866213626

(the error above is not from this build, but errors are the same)

### Output of `pulumi about`

```
CLI
Version      3.139.0
Go Version   go1.23.3
Go Compiler  gc

Plugins
KIND      NAME          VERSION
resource  azure-native  2.72.0
resource  cloudflare    5.43.0
resource  command       1.0.1
resource  docker        4.5.7
language  nodejs        unknown
resource  vercel        1.14.3

Host
OS       Microsoft Windows 11 Pro
Version  10.0.26120 Build 26120
Arch     x86_64

This project is written in nodejs: executable='C:\Program Files\nodejs\node.exe' version='v20.18.0'

(... redacted)

Found no pending operations associated with signalco/next

Backend
Name           pulumi.com
URL            https://app.pulumi.com/AleksandarDev
User           AleksandarDev
Organizations  AleksandarDev, signalco
Token type     personal

Pulumi locates its logs in C:\Users\aleks\AppData\Local\Temp by default
warning: Failed to get information about the Pulumi program's dependencies: no package-lock.json or yarn.lock file found (searching upwards from D:\source\signalco\infra\apps\cloud-primary)
```

Same issue is observed on `ubuntu-latest` GitHub actions runners.

### Additional context

Regarding warning from `pulumi about` - I'm using `pnpm` in monorepo but everything is running fin, maybe `pulumi about` doesn't pick up pnpm lockfiles correctly.

Using Service Principal for Azure auth.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by danielrbradley** [comment]
Hi @AleksandarDev thanks for raising this with us. We're in the process of moving to a newer, more maintained auth library and it looks like something's been missed around the blob rework in https://github.com/pulumi/pulumi-azure-native/pull/3698

We'll need to investigate and work out how to resolve in the next release.

**Comment by AleksandarDev** [comment]
@danielrbradley Thanks for quick reply. 

I see that you are in process of migrating from custom implementation (https://github.com/pulumi/pulumi-azure-native/pull/3698) but wanted to let you know there is reproducable issue here.

The  workaround you suggested works. I'll leave it unchanged for now on my project so we can confirm it works in next versions. If you need more debug info to test this before next release, let me know and I can quickly do run with nightly/canary build to confirm (since I didn't provide minimal repro).

**Comment by thomas11** [comment]
Hi @AleksandarDev, which of the provider's [authentication methods](https://www.pulumi.com/registry/packages/azure-native/installation-configuration/#authentication-methods) does your failing program use?

Is it simple for you to test it locally using `az` CLI auth?

**Comment by thomas11** [comment]
According to Azure docs, writing blobs requires the [Storage Blob Data Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles/storage#storage-blob-data-contributor) role. Note that [Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles/privileged#contributor) doesn't have the `dataActions`.

That would also explain why our CI tests pass - the SP is "Storage Blob Data Owner".

It doesn't explain, however, why things are working with the previous authentication backend. Will need to research more.

**Comment by thomas11** [comment]
Ah, I see. The old implementation, for blobs specifically, doesn't use the standard auth token to access blobs at all. It only uses it to list shared access keys for the storage account, which it then uses to write blobs.

**Comment by AleksandarDev** [comment]
@thomas11 I can confirm issue is resolved with `2.73.1` 🥳. Thanks for fast response and great support 🖖
